src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -7,115 +7,114 @@
- */
-static OpBase *_chain_cp_and_filter(const ExecutionPlan *plan, OpBase *lhs, OpBase *rhs,
-									OpBase *filter) {
-	OpBase *cp = NewCartesianProductOp(plan);
-	ExecutionPlan_AddOp(cp, lhs);
-	ExecutionPlan_AddOp(cp, rhs);
-	ExecutionPlan_AddOp(filter, cp);
-	return filter;
+// Finds all the cartesian product's children which solve a specific filter entities.
+static OpBase **_find_entities_solving_branches(rax *entities, OpBase *cp) {
+	int entities_count = raxSize(entities);
+	OpBase **solving_branches = array_new(OpBase *, 1);
+	for(int i = 0; i < cp->childCount; i++) {
+		OpBase *branch = cp->children[i];
+		/* Locate references reduces the amount of entities upon each call
+		 * that partially solves the references. */
+		ExecutionPlan_LocateReferences(branch, NULL, entities);
+		int new_entities_count = raxSize(entities);
+		if(new_entities_count != entities_count) {
+			// Update entity count.
+			entities_count = new_entities_count;
+			// Add partially solving branch into the array.
+			solving_branches = array_append(solving_branches, branch);
+			// No need to continue from here.
+			if(entities_count == 0) break;
+		}
+	}
+	assert(entities_count == 0);
+	assert(array_len(solving_branches) > 0);
+	// The rax containing the entities is now emtpy, free it.
+	raxFree(entities);
+	return solving_branches;
 }
 
 static void _optimize_cartesian_product(ExecutionPlan *plan, OpBase *cp) {
 	// Retrieve all filter operations located upstream from the Cartesian Product.
-	OpFilter **filter_ops = _locate_filters(cp);
-	uint filter_count = array_len(filter_ops);
-
-	// For each stream joined by the Cartesian product, collect all entities the stream resolves.
-	int stream_count = cp->childCount;
-	rax *stream_entities[stream_count];
-	OptimizeUtils_BuildStreamFromOp(cp, stream_entities, stream_count);
+	Filter_And_Entities *filters = _locate_filters_and_entities(cp);
+	uint filter_count = array_len(filters);
 
 	for(uint i = 0; i < filter_count; i ++) {
-		// Try to create a new dual-branched cartesian product, followed by the current filter.
-		OpFilter *filter_op = filter_ops[i];
-
-		/* A new cartesian product can be generated if both sides of the filter can be fully and
-		 * separately resolved by exactly two child streams. */
-		FT_FilterNode *f = filter_op->filterTree;
-
-		/* Make sure LHS of the filter is resolved by a stream. */
-		AR_ExpNode *lhs = f->pred.lhs;
-		uint lhs_resolving_stream = OptimizeUtils_RelateExpToStream(lhs, stream_entities, stream_count);
-		if(lhs_resolving_stream == NOT_RESOLVED) continue;
-		/* Make sure RHS of the filter is resolved by a stream. */
-		AR_ExpNode *rhs = f->pred.rhs;
-		uint rhs_resolving_stream = OptimizeUtils_RelateExpToStream(rhs, stream_entities, stream_count);
-		if(rhs_resolving_stream == NOT_RESOLVED) continue;
+		// Try to create a cartesian product, followed by the current filter.
+		OpFilter *filter_op = filters[i].filter;
+		ExecutionPlan_RemoveOp(plan, (OpBase *)filter_op);
```

Comment:
It seems like we're always going to reposition a filter
Consider: `MATCH (a), (b) WHERE a.v = b.v RETURN a,b`
the filter is already at the right position, I think we're about to create a new unnecessary CP and relocate the current filter beneath it.

Reply:
I changed the code to continue the filter loop in case of `solving_branch_count == cp->childCount`